lib/install/agent.go

```diff
@@ -18,99 +18,107 @@ package install
 
 import (
 	"context"
-	"fmt"
 	"net"
-	"strconv"
-	"strings"
+	"net/url"
 
 	"github.com/gravitational/gravity/lib/defaults"
 	"github.com/gravitational/gravity/lib/httplib"
-	"github.com/gravitational/gravity/lib/pack/webpack"
 	pb "github.com/gravitational/gravity/lib/rpc/proto"
 	rpcserver "github.com/gravitational/gravity/lib/rpc/server"
 	"github.com/gravitational/gravity/lib/utils"
 
-	"github.com/gravitational/roundtrip"
 	"github.com/gravitational/trace"
 	log "github.com/sirupsen/logrus"
 )
 
 // NewAgent returns a new unstarted agent instance
-func NewAgent(ctx context.Context, config AgentConfig, log log.FieldLogger, watchCh chan rpcserver.WatchEvent) (rpcserver.Server, error) {
+func NewAgent(ctx context.Context, config AgentConfig) (*rpcserver.PeerServer, error) {
 	if err := config.CheckAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
-
-	addr := config.PackageAddr
-	// Assume addr to be a complete address if it's prefixed with `http`
-	if !strings.Contains(addr, "http") {
-		host, port := utils.SplitHostPort(addr, strconv.Itoa(defaults.GravitySiteNodePort))
-		addr = fmt.Sprintf("https://%v:%v", host, port)
-	}
-
-	httpClient := roundtrip.HTTPClient(httplib.GetClient(true))
-	packages, err := webpack.NewBearerClient(addr, config.Token, httpClient)
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-
-	creds, err := LoadRPCCredentials(ctx, packages, log)
-	if err != nil {
+	if err := FetchCloudMetadata(config.CloudProvider, &config.RuntimeConfig); err != nil {
 		return nil, trace.Wrap(err)
 	}
-
 	listener, err := net.Listen("tcp", defaults.GravityRPCAgentAddr(config.AdvertiseAddr))
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-
+	defer func() {
+		if err != nil {
+			listener.Close()
+		}
+	}()
 	peerConfig := rpcserver.PeerConfig{
 		Config: rpcserver.Config{
-			Listener:      listener,
-			Credentials:   *creds,
-			RuntimeConfig: config.RuntimeConfig,
-		},
-		WatchCh: watchCh,
-		ReconnectStrategy: rpcserver.ReconnectStrategy{
-			ShouldReconnect: utils.ShouldReconnectPeer,
+			FieldLogger:     config.FieldLogger,
+			Listener:        listener,
+			Credentials:     config.Credentials,
+			RuntimeConfig:   config.RuntimeConfig,
+			AbortHandler:    config.AbortHandler,
+			CompleteHandler: config.CompleteHandler,
 		},
+		WatchCh:           config.WatchCh,
+		ReconnectStrategy: *config.ReconnectStrategy,
 	}
-
-	agent, err := rpcserver.NewPeer(peerConfig, config.ServerAddr, log)
+	agent, err := rpcserver.NewPeer(peerConfig, config.ServerAddr)
 	if err != nil {
 		listener.Close()
 		return nil, trace.Wrap(err)
 	}
-
+	if !config.SkipConnectValidation {
```

Comment:
Added this kludge as installer runs an agent which might start before the agent server is up.

Reply:
It does in general, but not this particular call which fails on first failure. I will disable it since I don't see value in this check.